### PR TITLE
Make sure 'final' decls never get marked as 'open'

### PR DESCRIPTION
**Pull request by jrose-apple**
Follow-up to #15996:

- Mark `willSet`/`didSet` accessors as `private`, always. These functions are always immediately inlined into a synthesized setter, and cannot be used with computed properties, so there's no benefit (and some harm) in having their access follow the property's.

- Disallow `open let`, and add a verifier check to make sure we don't mix `open` and `final` in the future. You could *write* `open let` in Swift 4.1, but you couldn't actually override such a property (within the module or outside). To maintain source compatibility, this is just going to emit a warning before Swift 5 (instead of an error), since it wasn't harming anything in practice.

I wouldn't have bothered with the first part right now except that it was the only other place where we ended up with declarations that were both `final` and `open` under the old code.

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux

**Review by slavapestov — APPROVED** [review]
Thanks for tightening up invariants!

**Comment by jrose-apple** [comment]
Hm, looks like I have more work to do!

**Comment by jrose-apple** [comment]
Two additional commits. @slavapestov, mind glancing it over once more?

@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -1223,6 +1224,57 @@ void swift::makeDynamic(ASTContext &ctx, ValueDecl *D) {
...
+  // Are there any reasons to infer 'final'? Prefer 'static' over the class
+  // being final for the purposes of diagnostics.
+  Optional<ImplicitlyFinalReason> reason;
+  if (staticSpelling == StaticSpellingKind::KeywordStatic) {
+    reason = ImplicitlyFinalReason::Static;
+
+    if (auto finalAttr = D->getAttrs().getAttribute<FinalAttr>()) {
+      auto finalRange = finalAttr->getRange();
+      if (finalRange.isValid()) {
+        TC.diagnose(finalRange.Start, diag::static_decl_already_final)
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: I think we have a diagnoseAndRemoveAttr() utility somewhere

**Review by slavapestov — APPROVED** [review]

File: lib/Sema/TypeCheckDecl.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckDecl.cpp]: So we do…but currently only in TypeCheckAttr.cpp. I think I've done enough moving little things around for the moment.

Comment (slavapestov) [reply lib/Sema/TypeCheckDecl.cpp]: No worries. There are many yaks to shave here and the PR is already a solid improvement.

**Pull request merged**